rpcs3/Emu/Io/usio.cpp

```diff
@@ -221,89 +222,162 @@ void usb_device_usio::translate_input()
+					if (pressed)
+						digital_input |= 0x800000ULL << shift;
 					break;
-				case usio_btn::strong_hit_side_left:
-					// Strong hit side left
+				case usio_btn::up:
 					if (pressed)
-						std::memcpy(input_buf.data() + 32 + offset, &c_big_hit, sizeof(u16));
+						digital_input |= 0x200000ULL << shift;
+					break;
+				case usio_btn::down:
+					if ( pressed)
+						digital_input |= 0x100000ULL << shift;
 					break;
-				case usio_btn::strong_hit_center_right:
-					// Strong hit center right
+				case usio_btn::left:
 					if (pressed)
-						std::memcpy(input_buf.data() + 36 + offset, &c_big_hit, sizeof(u16));
+						digital_input |= 0x80000ULL << shift;
 					break;
-				case usio_btn::strong_hit_side_right:
-					// Strong hit side right
+				case usio_btn::right:
 					if (pressed)
-						std::memcpy(input_buf.data() + 38 + offset, &c_big_hit, sizeof(u16));
+						digital_input |= 0x40000ULL << shift;
 					break;
-				case usio_btn::strong_hit_center_left:
-					// Strong hit center left
+				case usio_btn::tekken_button1:
 					if (pressed)
-						std::memcpy(input_buf.data() + 34 + offset, &c_big_hit, sizeof(u16));
+						digital_input |= 0x20000ULL << shift;
 					break;
-				case usio_btn::small_hit_center_left:
-					// Small hit center left
+				case usio_btn::tekken_button2:
 					if (pressed)
-						std::memcpy(input_buf.data() + 34 + offset, &c_small_hit, sizeof(u16));
+						digital_input |= 0x10000ULL << shift;
 					break;
-				case usio_btn::small_hit_center_right:
-					// Small hit center right
+				case usio_btn::tekken_button3:
 					if (pressed)
-						std::memcpy(input_buf.data() + 36 + offset, &c_small_hit, sizeof(u16));
+						digital_input |= 0x40000000ULL << shift;
 					break;
-				case usio_btn::small_hit_side_left:
-					// Small hit side left
+				case usio_btn::tekken_button4:
 					if (pressed)
-						std::memcpy(input_buf.data() + 32 + offset, &c_small_hit, sizeof(u16));
+						digital_input |= 0x20000000ULL << shift;
 					break;
-				case usio_btn::small_hit_side_right:
-					// Small hit side right
+				case usio_btn::tekken_button5:
 					if (pressed)
-						std::memcpy(input_buf.data() + 38 + offset, &c_small_hit, sizeof(u16));
+						digital_input |= 0x80000000ULL << shift;
 					break;
-				case usio_btn::count:
+				default:
 					break;
 				}
 			});
 	};
 
 	for (usz i = 0; i < g_cfg_usio.players.size(); i++)
-	{
 		translate_from_pad(i, i);
-	}
 
-	test_keys |= test_on ? 0x80 : 0x00;
-	std::memcpy(input_buf.data(), &test_keys, sizeof(u16));
-	std::memcpy(input_buf.data() + 16, &coin_counter, sizeof(u16));
+	digital_input |= test_on ? 0x80 : 0x00;
+	std::memcpy(input_buf.data() + 128, &digital_input, sizeof(u64));
+	std::memcpy(input_buf.data() + 128 + 16, &coin_counter, sizeof(u16));
+
+	constexpr std::array<bool, 8> dip_switches {false, false, false, false, true, false, false, false};
+	for (usz i = 0; i < dip_switches.size(); i++)
+		input_buf[2] |= dip_switches[i] << i;
```

Comment:
you can actually do this whole thing as constexpr by just doing:
```suggestion
		constexpr u32 dip_switches = 0b00010000;
		input_buf[2] |= dip_switches;
```
or if you want to keep the code:
```suggestion
	constexpr u32 dip_switches = []()
	{
		constexpr std::array<bool, 8> dip_switches {false, false, false, false, true, false, false, false};
		u32 value = 0;
		for (usz i = 0; i < dip_switches.size(); i++)
			value |= dip_switches[i] << i;
		return value;
	}();
	input_buf[2] |= dip_switches;
```

Reply:
That's nice.